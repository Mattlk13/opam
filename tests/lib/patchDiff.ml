type content =
  | File of string (* file *)
  | Dir of (string * string) list (* directory with list filename * content *)
  | Symlink (* Soft Link *)
  | Hardlink (* Hard link *)
  | V (* void *)

type arborescence = {
  name: string;
  first : content;
  second : content;
}

(** Contents *)

let foo = "foo\n"
let bar = "bar\n"
let foobar = "foo\nbar\n"
let same_file = {
  name = "same-file";
  first = File foo;
  second = File foo;
}
let diff_file = {
  name = "diff-file";
  first = File foo;
  second = File bar;
}
let diff_file_plus_fst = {
  name = "diff-file-plus-fst";
  first = File foobar;
  second = File foo;
}
let diff_file_plus_snd = {
  name = "diff-file-plus-snd";
  first = File foo;
  second = File foobar;
}

let content_working_diff = [
  same_file;
  diff_file;
  diff_file_plus_fst;
  diff_file_plus_snd;
  { name = "diff-file";
    first = File foo;
    second = File bar;
  };
  { name = "diff-file-plus-fst";
    first = File foobar;
    second = File foo;
  };
  { name = "diff-file-plus-snd";
    first = File foo;
    second = File foobar;
  };
  { name = "file-only-fst";
    first = File foo;
    second = V;
  };
  { name = "file-only-snd";
    first = V;
    second = File foo;
  };
  { name = "same-dir";
    first = Dir [];
    second = Dir [];
  };
  { name = "diff-dir-plus-fst";
    first = Dir [ "fst", foo ];
    second = Dir [ "fst", foobar ] ;
  };
  { name = "diff-dir-plus-snd";
    first = Dir [ "fst", foobar ];
    second = Dir [ "fst", foo ];
  };
  { name = "dir-only-fst";
    first = Dir [ "fst", foo ];
    second = V;
  };
]

let content_dir_file = [
  same_file;
  { name = "file-fst-dir-snd";
    first = File foo;
    second = Dir [ "fst", foo];
  };
]

let content_file_dir = [
  same_file;
  { name = "dir-fst-file-snd";
    first = Dir [ "fst", foo ];
    second = File foo;
  };
]

let content_symlink_fst = [
  same_file;
  { name = "linked-file-fst";
    first = Symlink;
    second = File foo;
  };
]

let content_symlink_snd = [
  same_file;
  { name = "linked-file-snd";
    first = File foo;
    second = Symlink;
  };
]

let content_hardlink_fst = [
  same_file;
  { name = "hardlinked-file-fst";
    first = Hardlink;
    second = File foo;
  };
]

let content_hardlink_snd = [
  same_file;
  { name = "hardlinked-file-snd";
    first = File foo;
    second = Hardlink;
  };
]


let content_patch_failure_garbage = [
  same_file;
  diff_file;
]
let diff_patch_failure_garbage =
  "something in\n" ^
  "the file\n" ^
  "that is not\n" ^
  "patch format\n"

let content_patch_failure_truncated = [
  same_file;
  diff_file;
  diff_file_plus_fst;
]
let diff_patch_failure_truncated =
  "--- first/diff-file\n" ^
  "+++ second/diff-file\n" ^
  "@@ -1,1 +1,1 @@\n" ^
  "-foo\n" ^
  "+bar\n" ^
  "--- first/diff-fi\n"

let _good_diff =
  "\n" ^
  "--- first/diff-file\n" ^
  "+++ second/diff-file\n" ^
  "@@ -1,1 +1,1 @@\n" ^
  "-foo\n" ^
  "+bar\n" ^
  "--- first/diff-file-plus-fst\n" ^
  "+++ second/diff-file-plus-fst\n" ^
  "@@ -2,1 +2,0 @@\n" ^
  "-bar\n"

(** Utils *)

let print = Printf.printf

open OpamFilename.Op
let read_dir root names =
  let lst =
    List.map (fun name ->
        let dir = root / name in
        (name^"/", []) ::
        List.map (fun f ->
            OpamFilename.remove_prefix root f,
            OpamStd.String.split (OpamFilename.read f) '\n')
          (OpamFilename.rec_files dir)
        @
        List.map (fun d -> OpamFilename.remove_prefix_dir root d, [])
          (OpamFilename.rec_dirs dir))
      names
    |> List.flatten
    |> List.map (fun (file, content) ->
        (OpamSystem.back_to_forward file, content))
  in

  let lst = List.sort (fun (f,_) (f',_) -> String.compare f f') lst in
  OpamStd.Format.itemize ~bullet:"+ "
    (function
      | d, [] -> d
      | d, c ->
        Printf.sprintf "%s\n%s"
          d
          (OpamStd.List.concat_map ~left:"" ~right:"" "\n"
             (Printf.sprintf "> %s") c))
    lst

let first = "first"
let second = "second"

let write_setup dir content =
  let first_root = dir / first in
  let second_root = dir / second in
  List.iter (fun d ->
      OpamFilename.cleandir d;
      OpamFilename.mkdir d)
    [ first_root; second_root; ] ;
  let link_f =
    let link = lazy (
      let f = dir // "linked_file" in
      if not (OpamFilename.exists f) then
        OpamFilename.write f bar;
      f
    ) in
    fun () -> Lazy.force link
  in
  let create inner_dir name = function
    | File content ->
      OpamFilename.write (inner_dir // name) content
    | Dir lst ->
      let inner_dir = inner_dir / name in
      List.iter (fun (n,c) -> OpamFilename.write (inner_dir // n) c) lst
    | Symlink ->
      OpamFilename.link ~relative:false ~target:(link_f ())
        ~link:(inner_dir // name)
    | Hardlink ->
      let target = OpamFilename.to_string (link_f ()) in
      let link = OpamFilename.to_string (inner_dir // name) in
      Unix.link target link
    | V -> ()
  in
  List.iter (fun {name; first; second} ->
      create first_root name first;
      create second_root name second)
    content

type diff_patch =
  | DiffPatch
  | Patch of string

let print_dirs dir =
  print "%s\n" (read_dir dir [ first; second ])

let diff_patch dir content kind =
  write_setup dir content;
  print "*** SETUP ***\n";
  print_dirs dir;
  let diff =
    match kind with
    | Patch patch ->
      print "*** GIVEN DIFF ***\n";
      let fpatch = dir // "patch" in
      OpamFilename.write fpatch patch;
      Some fpatch
    | DiffPatch ->
      print "*** DIFF ***\n";
      match
        OpamRepositoryBackend.get_diff dir
          (OpamFilename.Base.of_string first)
          (OpamFilename.Base.of_string second)
      with
      | exception Failure s -> print "ERROR: %s\n" s; None
      | None -> print "No diff\n"; None
      | some -> some
  in
  match diff with
  | None -> ()
  | Some diff ->
    print "%s\n" (OpamFilename.read diff);
    let result =
      OpamFilename.patch ~allow_unclean:false diff
        (dir / first)
    in
    match result with
    | None ->
      print "*** PATCHED ***\n";
      print_dirs dir
    | Some exn ->
      print "*** PATCH ERROR ***\n";
      print "ERROR: %s\n" (Printexc.to_string exn)

(** The tests *)

let tests = [
  "normal", content_working_diff, DiffPatch;
  "diff file/dir error", content_dir_file, DiffPatch;
  "diff dir/file error", content_file_dir, DiffPatch;
  "symlink fst", content_symlink_fst, DiffPatch;
  "symlink snd", content_symlink_snd, DiffPatch;
  "hardlink fst", content_hardlink_fst, DiffPatch;
  "hardlink snd", content_hardlink_snd, DiffPatch;
  "patch error garbage", content_patch_failure_garbage,
  Patch diff_patch_failure_garbage;
  "patch truncated", content_patch_failure_truncated,
  Patch diff_patch_failure_truncated;
]

let () =
  (* This causes Windows to use LF endings instead of CRLF, which simplifies the comparison with the reference file *)
  Unix.putenv "LC_ALL" "C";
  set_binary_mode_out stdout true;
  Unix.dup2 Unix.stdout Unix.stderr;
  OpamFilename.with_tmp_dir @@ fun dir ->
  List.iteri (fun i (label, content, kind) ->
      print "\n----------------------\n";
      print " Test %d: %s\n" (i+1) label;
      print "----------------------\n\n";
      diff_patch dir content kind)
    tests
