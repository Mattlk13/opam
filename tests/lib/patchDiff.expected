
----------------------
 Test 1: normal
----------------------

*** SETUP ***
+ first/
+ first/diff-dir-plus-fst
+ first/diff-dir-plus-fst/fst
  > foo
+ first/diff-dir-plus-snd
+ first/diff-dir-plus-snd/fst
  > foo
  > bar
+ first/diff-file
  > foo
+ first/diff-file-plus-fst
  > foo
  > bar
+ first/diff-file-plus-snd
  > foo
+ first/dir-only-fst
+ first/dir-only-fst/fst
  > foo
+ first/file-only-fst
  > foo
+ first/same-dir
+ first/same-file
  > foo
+ second/
+ second/diff-dir-plus-fst
+ second/diff-dir-plus-fst/fst
  > foo
  > bar
+ second/diff-dir-plus-snd
+ second/diff-dir-plus-snd/fst
  > foo
+ second/diff-file
  > bar
+ second/diff-file-plus-fst
  > foo
+ second/diff-file-plus-snd
  > foo
  > bar
+ second/file-only-snd
  > foo
+ second/same-dir
+ second/same-file
  > foo

*** DIFF ***
--- first/diff-dir-plus-fst/fst
+++ second/diff-dir-plus-fst/fst
@@ -2,0 +2,1 @@
+bar
--- first/diff-dir-plus-snd/fst
+++ second/diff-dir-plus-snd/fst
@@ -2,1 +2,0 @@
-bar
--- first/diff-file
+++ second/diff-file
@@ -1,1 +1,1 @@
-foo
+bar
--- first/diff-file-plus-fst
+++ second/diff-file-plus-fst
@@ -2,1 +2,0 @@
-bar
--- first/diff-file-plus-snd
+++ second/diff-file-plus-snd
@@ -2,0 +2,1 @@
+bar
--- first/dir-only-fst/fst
+++ /dev/null
@@ -1,1 +0,0 @@
-foo
--- first/file-only-fst
+++ /dev/null
@@ -1,1 +0,0 @@
-foo
--- /dev/null
+++ second/file-only-snd
@@ -0,0 +1,1 @@
+foo

*** PATCHED ***
+ first/
+ first/diff-dir-plus-fst
+ first/diff-dir-plus-fst/fst
  > foo
  > bar
+ first/diff-dir-plus-snd
+ first/diff-dir-plus-snd/fst
  > foo
+ first/diff-file
  > bar
+ first/diff-file-plus-fst
  > foo
+ first/diff-file-plus-snd
  > foo
  > bar
+ first/dir-only-fst
+ first/file-only-snd
  > foo
+ first/same-dir
+ first/same-file
  > foo
+ second/
+ second/diff-dir-plus-fst
+ second/diff-dir-plus-fst/fst
  > foo
  > bar
+ second/diff-dir-plus-snd
+ second/diff-dir-plus-snd/fst
  > foo
+ second/diff-file
  > bar
+ second/diff-file-plus-fst
  > foo
+ second/diff-file-plus-snd
  > foo
  > bar
+ second/file-only-snd
  > foo
+ second/same-dir
+ second/same-file
  > foo

*** GIT DIFF ***
diff --git b/diff-dir-plus-fst/fst a/diff-dir-plus-fst/fst
index c0ffee..c0ffee c0ffee
--- b/diff-dir-plus-fst/fst
+++ a/diff-dir-plus-fst/fst
@@ -1 +1,2 @@
 foo
+bar
diff --git b/diff-dir-plus-snd/fst a/diff-dir-plus-snd/fst
index c0ffee..c0ffee c0ffee
--- b/diff-dir-plus-snd/fst
+++ a/diff-dir-plus-snd/fst
@@ -1,2 +1 @@
 foo
-bar
diff --git b/diff-file a/diff-file
index c0ffee..c0ffee c0ffee
--- b/diff-file
+++ a/diff-file
@@ -1 +1 @@
-foo
+bar
diff --git b/diff-file-plus-fst a/diff-file-plus-fst
index c0ffee..c0ffee c0ffee
--- b/diff-file-plus-fst
+++ a/diff-file-plus-fst
@@ -1,2 +1 @@
 foo
-bar
diff --git b/diff-file-plus-snd a/diff-file-plus-snd
index c0ffee..c0ffee c0ffee
--- b/diff-file-plus-snd
+++ a/diff-file-plus-snd
@@ -1 +1,2 @@
 foo
+bar
diff --git b/file-only-fst a/file-only-fst
deleted file mode c0ffee
index c0ffee..c0ffee
--- b/file-only-fst
+++ /dev/null
@@ -1 +0,0 @@
-foo
diff --git b/dir-only-fst/fst a/file-only-snd
similarity index 100%
rename from dir-only-fst/fst
rename to file-only-snd

*** GIT PATCHED ***
+ first/
+ first/diff-dir-plus-fst
+ first/diff-dir-plus-fst/fst
  > foo
  > bar
+ first/diff-dir-plus-snd
+ first/diff-dir-plus-snd/fst
  > foo
+ first/diff-file
  > bar
+ first/diff-file-plus-fst
  > foo
+ first/diff-file-plus-snd
  > foo
  > bar
+ first/dir-only-fst
+ first/file-only-snd
  > foo
+ first/same-dir
+ first/same-file
  > foo
+ second/
+ second/diff-dir-plus-fst
+ second/diff-dir-plus-fst/fst
  > foo
  > bar
+ second/diff-dir-plus-snd
+ second/diff-dir-plus-snd/fst
  > foo
+ second/diff-file
  > bar
+ second/diff-file-plus-fst
  > foo
+ second/diff-file-plus-snd
  > foo
  > bar
+ second/file-only-snd
  > foo
+ second/same-dir
+ second/same-file
  > foo


----------------------
 Test 2: diff file/dir error
----------------------

*** SETUP ***
+ first/
+ first/file-fst-dir-snd
  > foo
+ first/same-file
  > foo
+ second/
+ second/file-fst-dir-snd
+ second/file-fst-dir-snd/fst
  > foo
+ second/same-file
  > foo

*** DIFF ***
ERROR: Change from a regular file to a directory is unsupported

----------------------
 Test 3: diff dir/file error
----------------------

*** SETUP ***
+ first/
+ first/dir-fst-file-snd
+ first/dir-fst-file-snd/fst
  > foo
+ first/same-file
  > foo
+ second/
+ second/dir-fst-file-snd
  > foo
+ second/same-file
  > foo

*** DIFF ***
ERROR: Change from a directory to a regular file is unsupported

----------------------
 Test 4: symlink fst
----------------------

*** SETUP ***
+ first/
+ first/linked-file-fst
  > bar
+ first/same-file
  > foo
+ second/
+ second/linked-file-fst
  > foo
+ second/same-file
  > foo

*** DIFF ***
ERROR: Symlinks are unsupported

----------------------
 Test 5: symlink snd
----------------------

*** SETUP ***
+ first/
+ first/linked-file-snd
  > foo
+ first/same-file
  > foo
+ second/
+ second/linked-file-snd
  > bar
+ second/same-file
  > foo

*** DIFF ***
ERROR: Symlinks are unsupported

----------------------
 Test 6: hardlink fst
----------------------

*** SETUP ***
+ first/
+ first/hardlinked-file-fst
  > bar
+ first/same-file
  > foo
+ second/
+ second/hardlinked-file-fst
  > foo
+ second/same-file
  > foo

*** DIFF ***
--- first/hardlinked-file-fst
+++ second/hardlinked-file-fst
@@ -1,1 +1,1 @@
-bar
+foo

*** PATCHED ***
+ first/
+ first/hardlinked-file-fst
  > foo
+ first/same-file
  > foo
+ second/
+ second/hardlinked-file-fst
  > foo
+ second/same-file
  > foo


----------------------
 Test 7: hardlink snd
----------------------

*** SETUP ***
+ first/
+ first/hardlinked-file-snd
  > foo
+ first/same-file
  > foo
+ second/
+ second/hardlinked-file-snd
  > foo
+ second/same-file
  > foo

*** DIFF ***
No diff

----------------------
 Test 8: patch error garbage
----------------------

*** SETUP ***
+ first/
+ first/diff-file
  > foo
+ first/same-file
  > foo
+ second/
+ second/diff-file
  > bar
+ second/same-file
  > foo

*** GIVEN DIFF ***
something in
the file
that is not
patch format

*** PATCHED ***
+ first/
+ first/diff-file
  > foo
+ first/same-file
  > foo
+ second/
+ second/diff-file
  > bar
+ second/same-file
  > foo


----------------------
 Test 9: patch truncated
----------------------

*** SETUP ***
+ first/
+ first/diff-file
  > foo
+ first/diff-file-plus-fst
  > foo
  > bar
+ first/same-file
  > foo
+ second/
+ second/diff-file
  > bar
+ second/diff-file-plus-fst
  > foo
+ second/same-file
  > foo

*** GIVEN DIFF ***
--- first/diff-file
+++ second/diff-file
@@ -1,1 +1,1 @@
-foo
+bar
--- first/diff-fi

*** PATCHED ***
+ first/
+ first/diff-file
  > bar
+ first/diff-file-plus-fst
  > foo
  > bar
+ first/same-file
  > foo
+ second/
+ second/diff-file
  > bar
+ second/diff-file-plus-fst
  > foo
+ second/same-file
  > foo


----------------------
 Test 10: add empty file
----------------------

*** SETUP ***
+ first/
+ first/same-file
  > foo
+ second/
+ second/im-empty
+ second/same-file
  > foo

*** DIFF ***
diff --git second/im-empty second/im-empty
new file mode 100644

*** PATCHED ***
+ first/
+ first/im-empty
+ first/same-file
  > foo
+ second/
+ second/im-empty
+ second/same-file
  > foo

*** GIT DIFF ***
diff --git b/im-empty a/im-empty
new file mode c0ffee
index c0ffee..c0ffee

*** GIT PATCHED ***
+ first/
+ first/im-empty
+ first/same-file
  > foo
+ second/
+ second/im-empty
+ second/same-file
  > foo


----------------------
 Test 11: remove empty file
----------------------

*** SETUP ***
+ first/
+ first/im-empty
+ first/same-file
  > foo
+ second/
+ second/same-file
  > foo

*** DIFF ***
diff --git first/im-empty first/im-empty
deleted file mode 100644

*** PATCHED ***
+ first/
+ first/same-file
  > foo
+ second/
+ second/same-file
  > foo

*** GIT DIFF ***
diff --git b/im-empty a/im-empty
deleted file mode c0ffee
index c0ffee..c0ffee

*** GIT PATCHED ***
+ first/
+ first/same-file
  > foo
+ second/
+ second/same-file
  > foo


----------------------
 Test 12: move file into a new directory
----------------------

*** SETUP ***
+ first/
+ first/move-me
  > bar
+ first/same-file
  > foo
+ second/
+ second/inner
+ second/inner/move-me
  > bar
+ second/same-file
  > foo

*** DIFF ***
--- /dev/null
+++ second/inner/move-me
@@ -0,0 +1,1 @@
+bar
--- first/move-me
+++ /dev/null
@@ -1,1 +0,0 @@
-bar

*** PATCHED ***
+ first/
+ first/inner
+ first/inner/move-me
  > bar
+ first/same-file
  > foo
+ second/
+ second/inner
+ second/inner/move-me
  > bar
+ second/same-file
  > foo

*** GIT DIFF ***
diff --git b/move-me a/inner/move-me
similarity index 100%
rename from move-me
rename to inner/move-me

*** GIT PATCH ERROR ***
ERROR: Unix.Unix_error(Unix.ENOENT, "rename", "/tmp/build_c0ffee_dune/opam-c0ffee-c0ffee/first/move-me")

----------------------
 Test 13: delete file that deletes the directory
----------------------

*** SETUP ***
+ first/
+ first/im-here
+ first/im-here/delete-me
  > bar
+ first/im-not-here
+ first/im-not-here/delete-me
  > baz
+ first/same-file
  > foo
+ second/
+ second/im-here
+ second/same-file
  > foo

*** DIFF ***
--- first/im-here/delete-me
+++ /dev/null
@@ -1,1 +0,0 @@
-bar
--- first/im-not-here/delete-me
+++ /dev/null
@@ -1,1 +0,0 @@
-baz
\ No newline at end of file

*** PATCHED ***
+ first/
+ first/im-here
+ first/im-not-here
+ first/same-file
  > foo
+ second/
+ second/im-here
+ second/same-file
  > foo

*** GIT DIFF ***
diff --git b/im-here/delete-me a/im-here/delete-me
deleted file mode c0ffee
index c0ffee..c0ffee
--- b/im-here/delete-me
+++ /dev/null
@@ -1 +0,0 @@
-bar
diff --git b/im-not-here/delete-me a/im-not-here/delete-me
deleted file mode c0ffee
index c0ffee..c0ffee
--- b/im-not-here/delete-me
+++ /dev/null
@@ -1 +0,0 @@
-baz
\ No newline at end of file

*** GIT PATCHED ***
+ first/
+ first/im-here
+ first/im-not-here
+ first/same-file
  > foo
+ second/
+ second/im-here
+ second/same-file
  > foo

